README.md
# Sinatra ActiveRecord Extension

Extends [Sinatra](http://www.sinatrarb.com/) with extension methods and Rake
tasks for dealing with an SQL database using the
[ActiveRecord ORM](https://github.com/rails/rails/tree/master/activerecord).

## Setup

Put it in your `Gemfile`, along with the adapter of your database. For
simplicity, let's assume you're using SQLite:

```ruby
gem "sinatra-activerecord"
gem "sqlite3"
gem "rake"
```

Now require it in your Sinatra application, and establish the database
connection:

```ruby
# app.rb
require "sinatra/activerecord"

set :database, {adapter: "sqlite3", database: "foo.sqlite3"}
# or set :database_file, "path/to/database.yml"
```

If you have a `config/database.yml`, it will automatically be loaded, no need
to specify it. Also, in production, the `$DATABASE_URL` environment variable
will automatically be read as the database (if you haven't specified otherwise).

Note that in **modular** Sinatra applications you will need to first register
the extension:

```ruby
class YourApplication < Sinatra::Base
  register Sinatra::ActiveRecordExtension
end
```

Now require the rake tasks and your app in your `Rakefile`:

```ruby
# Rakefile
require "sinatra/activerecord/rake"

namespace :db do
  task :load_config do
    require "./app"
  end
end
```

In case you're using tools like SimpleCov, where the application needs to be
required only inside SimpleCov's Rake tasks, you can do

```ruby
# Rakefile
require "sinatra/activerecord/rake"

namespace :db do
  task :load_config do
    require "./app"
  end
end
```

In the Terminal test that it works:

```sh
$ bundle exec rake -T
rake db:create            # Create the database from DATABASE_URL or config/database.yml for the current Rails.env (use db:create:all to create all dbs in the config)
rake db:create_migration  # Create a migration (parameters: NAME, VERSION)
rake db:drop              # Drops the database using DATABASE_URL or the current Rails.env (use db:drop:all to drop all databases)
rake db:fixtures:load     # Load fixtures into the current environment's database
rake db:migrate           # Migrate the database (options: VERSION=x, VERBOSE=false)
rake db:migrate:status    # Display status of migrations
rake db:rollback          # Rolls the schema back to the previous version (specify steps w/ STEP=n)
rake db:schema:dump       # Create a db/schema.rb file that can be portably used against any DB supported by AR
rake db:schema:load       # Load a schema.rb file into the database
rake db:seed              # Load the seed data from db/seeds.rb
rake db:setup             # Create the database, load the schema, and initialize with the seed data (use db:reset to also drop the db first)
rake db:structure:dump    # Dump the database structure to db/structure.sql
rake db:version           # Retrieves the current schema version number
```

And that's it, you're all set :)

## Usage

You can create a migration:

```sh
$ bundle exec rake db:create_migration NAME=create_users
```

This will create a migration file in your migrations directory (`./db/migrate`
by default), ready for editing.

```ruby
class CreateUsers < ActiveRecord::Migration
  def change
    create_table :users do |t|
      t.string :name
    end
  end
end
```

Now migrate the database:

```sh
$ bundle exec rake db:migrate
```

You can also write models:

```ruby
class User < ActiveRecord::Base
  validates_presence_of :name
end
```

You can put your models anywhere you want, only remember to require them if
they're in a separate file, and that they're loaded after `require "sinatra/activerecord"`.

Now everything just works:

```ruby
get '/users' do
  @users = User.all
  erb :index
end

get '/users/:id' do
  @user = User.find(params[:id])
  erb :show
end
```

A nice thing is that the `ActiveRecord::Base` class is available to
you through the `database` variable:

```ruby
if database.table_exists?('users')
  # Do stuff
else
  raise "The table 'users' doesn't exist."
end
```

## History

This gem was made in 2009 by Blake Mizerany, creator of Sinatra.

## Social

You can follow me on Twitter, I'm [@jankomarohnic](http://twitter.com/jankomarohnic).

## License

[MIT](https://github.com/janko-m/sinatra-activerecord/blob/master/LICENSE)


CHANGELOG.md
# Changelog

## Version 2.0.25
- Added support for ActiveRecord 7.0

## Version 2.0.24
- Fixed compatibility with Psych version 4.0.0+ , use unsafe_load if available.
- Added Ruby 3.0 and head (latest version of Ruby) to the CI matrix
Thanks Onaka (@onk) for contributing to this.

## Version 2.0.23
- For ActiveRecord 6.0 and above, if the the configuration (database.yml) has multiple databases defined, it will discard the DATABASE_URL spec (if it is supplied from environment variables), following [Active Record convention here](https://github.com/rails/rails/blob/main/activerecord/lib/active_record/database_configurations.rb#L169). Thanks JoakimKlaxit (@JoakimKlaxit) for contributing to this.


## Version 2.0.22

- Added compability fix for ActiveRecord 6.1 ([#107](https://github.com/janko-m/sinatra-activerecord/issue/107)), as they have moved ConnectionUrlResolver to another module. Thanks Richard Peck(@richpeck) for contributing to this.

## Version 2.0.21

- If both config/database.yml and $DATABASE_URL is present, the database configuration will be merged from this two, and $DATABASE_URL's variables will take precedence. (https://github.com/sinatra-activerecord/sinatra-activerecord/pull/103)

## Version 2.0.9

- Don't try to load config/database.yml if `$DATABASE_URL` is present ([#55](https://github.com/janko-m/sinatra-activerecord/pull/55), thanks to @exviva)

## Version 2.0.8

- Use the ConnectionManagement AR middelware for clearing active connections

## Version 2.0.7

- Allow multiple databases to be used ([#54](https://github.com/janko-m/sinatra-activerecord/pull/54))

## Version 2.0.6

* Prefer `DATABASE_URL` to config/database.yml ([#47](https://github.com/janko-m/sinatra-activerecord/issue/47)).

## Version 2.0.5

* Disabled ActiveRecord logging in tests

## Version 2.0.4

* You're now encouraged to load the app in the `:load_config` Rake task. This
  way your app will only be required when executing ActiveRecord tasks, and
  not other potential tasks which don't need the database. Your Rakefile can
  now be updated to this:

  ```rb
  require "sinatra/activerecord/rake"

  namespace :db do
    task :load_config do
      require "./app" # or whatever your app name is
    end
  end
  ```

* ActiveRecord isn't logged in production anymore.

## Version 2.0.3

- Use `File.exist?` instead of `File.exists?` to avoid warnings (thanks to [**@matthiase**](https://github.com/matthiase)).

## Version 2.0.2

- Allow setting the environment as a String, aside from Symbol.

## Version 2.0.1

- Fixed the issue where it mattered in which order `sinatra/activerecord/rake`
  and `./app` are required.

## Version 2.0.0

- ActiveRecord 4.1 is now supported.
- Rake tasks are now directly imported from ActiveRecord, which means *all* of
  the tasks are now available.

## Version 1.7.0

- Added `db:reset` and `db:migrate:reset` tasks ([#27](https://github.com/janko-m/sinatra-activerecord/pull/27)).
  Thanks to @csgavino.

## Version 1.6.0

- Added ability to specify folder for the database ([#26](https://github.com/janko-m/sinatra-activerecord/pull/26)).
  Thanks to @jamesrwhite.

## Version 1.5.0

- Added ability to specify `:sql` format of the schema, as well as
  `db:structure` tasks ([#25](https://github.com/janko-m/sinatra-activerecord/pull/25)).
  Thanks to @icambron.

## Version 1.4.0

- Added `db:test:prepare` Rake task.

## Version 1.3.0

- Added `db:create`, `db:drop` and `db:reset` Rake tasks.

## Version 1.2.5

- Fix a bug with ActiveRecord 3.0.

## Version 1.2.4

- By default look for a `config/database.yml` in the root of the running process.

## Version 1.2.3

- Allow ActiveRecord 4

## Version 1.2.2

- Fix not allowing environments other than "development", "test" and "production".

## Version 1.2.1

- Fix specifying database from a file not working properly.

## Version 1.2.0

- You can now set a custom migrations directory. Take a look at
[this wiki](https://github.com/janko-m/sinatra-activerecord/wiki/Changing-the-migrations-directory).

- You can now set the database to a YAML file. Take a look at
[this wiki](https://github.com/janko-m/sinatra-activerecord/wiki/Alternative-database-setup).

## Version 1.1.2

- Users are now warned if they use SQLite URLs with 2 slashes.

## Version 1.1.1

- The `set :database` command now accepts everything that
  `ActiveRecord::Base.establish_connection` accepts. A Hash
  is one example:

```ruby
set :database, {
  adapter: "sqlite3",
  database: "db/foo.db",
  pool: 5,
  timeout: 5000
}
```

- The database is now reset whenever you specify it with `set :database`.
- The migrations are now logged just like in Rails.

## Version 1.1.0

- If no database is specified, it will try to read from the `DATABASE_URL`
  environment variable (which is often set in production). This was
  removed in 1.0.0, but now I'm bringing it back.

## Version 1.0.1

- Removed deprecation warnings when using the `#database` helper.

## Version 1.0.0

- Fixed the database not working. Sorry, I won't let it happen again.

### Changes that are NOT backwards compatible

- There is no more "default" database. Now you always have to specify
  the database you want to connect to, it doesn't default to
  `"sqlite://#{environment}.db"` anymore.

- When you're setting an SQLite database, you now have to put 3 slashes
  after `sqlite:` instead of 2. So instead of `sqlite://database.db` you
  have to write `sqlite:///database.db`. This is now a valid URL.

## Version 0.2.1 (yanked)

- The previous version was yanked, because I forgot to add the
  `db:rollback` task (I didn't figure out how to test it yet, otherwise
  I would know it was missing). This version is then actually 0.2.0.

## Version 0.2.0 (yanked)

- Added `db:rollback` rake task.

- If you're using SQLite, you can now specify the path to where
  you want your database file to be (thanks to **@mpalmer** for this). Refer to
  [this wiki](https://github.com/janko-m/sinatra-activerecord/wiki/SQLite).


lib/sinatra/activerecord.rb
require 'sinatra/base'
require 'active_record'
require 'active_support/core_ext/hash/keys'

require 'logger'
require 'pathname'
require 'yaml'
require 'erb'

require 'active_record/database_configurations/connection_url_resolver' if Gem.loaded_specs["activerecord"].version >= Gem::Version.create('6.1')

module Sinatra
  module ActiveRecordHelper
    def database
      settings.database
    end
  end

  module ActiveRecordExtension
    def self.registered(app)
      if ENV['DATABASE_URL'] && File.exist?("#{Dir.pwd}/config/database.yml")
        path = "#{Dir.pwd}/config/database.yml"
        url = ENV['DATABASE_URL']
        file_path = File.join(root, path) if Pathname(path).relative? and root
        source = ERB.new(File.read(path)).result
        file_spec = YAML.respond_to?(:unsafe_load) ? YAML.unsafe_load(source) : YAML.load(source)
        file_spec ||= {}

        # ActiveRecord 6.1+ has moved the connection url resolver to another module
        if Gem.loaded_specs["activerecord"].version >= Gem::Version.create('6.1')
          url_spec = ActiveRecord::DatabaseConfigurations::ConnectionUrlResolver.new(url).to_hash
        else
          url_spec = ActiveRecord::ConnectionAdapters::ConnectionSpecification::ConnectionUrlResolver.new(url).to_hash
        end
        
        # if the configuration concerns only one database, and url_spec exist, url_spec will override the same key
        # if the configuration has multiple databases (Rails 6.0+ feature), url_spec is discarded
        # Following Active Record config convention
        # https://github.com/rails/rails/blob/main/activerecord/lib/active_record/database_configurations.rb#L169
        final_spec = file_spec.keys.map do |env|
          config = file_spec[env]
          if config.is_a?(Hash) && config.all? { |_k, v| v.is_a?(Hash) }
            [env, config]
          else
            [env, config.merge(url_spec)]
          end
        end.to_h

        app.set :database, final_spec
      elsif ENV['DATABASE_URL']
        app.set :database, ENV['DATABASE_URL']
      elsif File.exist?("#{Dir.pwd}/config/database.yml")
        app.set :database_file, "#{Dir.pwd}/config/database.yml"
      end

      unless defined?(Rake) || [:test, :production].include?(app.settings.environment)
        ActiveRecord::Base.logger = Logger.new(STDOUT)
      end

      app.helpers ActiveRecordHelper

      # re-connect if database connection dropped (Rails 3 only)
      app.before { ActiveRecord::Base.verify_active_connections! if ActiveRecord::Base.respond_to?(:verify_active_connections!) }
      app.after { ActiveRecord::Base.clear_active_connections! }
    end

    def database_file=(path)
      path = File.join(root, path) if Pathname(path).relative? and root
      source = ERB.new(File.read(path)).result
      spec = YAML.respond_to?(:unsafe_load) ? YAML.unsafe_load(source) : YAML.load(source)
      spec ||= {}
      set :database, spec
    end

    def database=(spec)
      if spec.is_a?(Hash) and spec.symbolize_keys[environment.to_sym]
        ActiveRecord::Base.configurations = spec.stringify_keys
        ActiveRecord::Base.establish_connection(environment.to_sym)
      elsif spec.is_a?(Hash)     
        ActiveRecord::Base.configurations = {
          environment.to_s => spec.stringify_keys
        }

        ActiveRecord::Base.establish_connection(spec.stringify_keys)
      else
        if Gem.loaded_specs["activerecord"].version >= Gem::Version.create('6.0')
          ActiveRecord::Base.configurations ||= ActiveRecord::DatabaseConfigurations.new({}).resolve(spec)
        else
          ActiveRecord::Base.configurations ||= {}
          ActiveRecord::Base.configurations[environment.to_s] = ActiveRecord::ConnectionAdapters::ConnectionSpecification::ConnectionUrlResolver.new(spec).to_hash
        end

        ActiveRecord::Base.establish_connection(spec)
      end
    end

    def database
      ActiveRecord::Base
    end
  end

  register ActiveRecordExtension
end
